src/protocol/boolean/solved_bits.rs

```diff
@@ -0,0 +1,265 @@
+
+        //
+        // step 3, 4 & 5
+        //
+        // if b >= p, then abort by returning `None`
+        if !Self::is_less_than_p(ctx.clone(), record_id, &b_b).await? {
+            return Ok(None);
+        }
+
+        //
+        // step 6
+        //
+        // if success, then compute `[b_p]` by `Σ 2^i * [b_i]_B`
+        #[allow(clippy::cast_possible_truncation)]
+        let b_p: Replicated<F> = b_b
+            .iter()
+            .enumerate()
+            .fold(Replicated::ZERO, |acc, (i, x)| {
+                acc + &(x.clone() * F::from(2_u128.pow(i as u32)))
+            });
+
+        Ok(Some(RandomBitsShare { b_b, b_p }))
+    }
+
+    /// Generates a sequence of `l` random bit sharings in the target field `F`.
+    async fn generate_random_bits<F: Field>(
+        ctx: ProtocolContext<'_, Replicated<F>, F>,
+        record_id: RecordId,
+    ) -> Result<Vec<Replicated<F>>, Error> {
+        // We assume the bit length we operate in would not exceed 255
+        #[allow(clippy::cast_possible_truncation)]
+        let l = F::Integer::BITS as u8;
+
+        // Generate a pair of random numbers. We'll use these numbers as
+        // the source of `l`-bit long uniformly random sequence of bits.
+        let (b_bits_left, b_bits_right) = ctx
+            .narrow(&Step::RandomFields)
+            .prss()
+            .generate_values(record_id);
+
+        // Same here. For now, 64-bit is enough for our F_p
+        #[allow(clippy::cast_possible_truncation)]
+        let xor_shares = XorShares::new(l, b_bits_left as u64, b_bits_right as u64);
+
+        // Convert each bit to secret sharings of that bit in the target field
+        let c = ctx.narrow(&Step::ConvertShares);
+        let futures = (0..l).map(|i| {
+            // again, we don't expect our prime field to be > 2^64
+            #[allow(clippy::cast_possible_truncation)]
+            let c = c.narrow(&BitOpStep::Step(i as usize));
+            async move {
+                ConvertShares::new(xor_shares)
+                    .execute_one_bit(c, record_id, i as u8)
+                    .await
+            }
+        });
+
+        try_join_all(futures).await
+    }
+
+    async fn is_less_than_p<F: Field>(
+        ctx: ProtocolContext<'_, Replicated<F>, F>,
+        record_id: RecordId,
+        b_b: &[Replicated<F>],
+    ) -> Result<bool, Error> {
+        let p_b = Self::local_secret_shared_bits_p(ctx.role());
+        let c_b =
+            BitwiseLessThan::execute(ctx.narrow(&Step::IsPLessThanB), record_id, b_b, &p_b).await?;
+        if ctx.narrow(&Step::RevealC).reveal(record_id, &c_b).await? == F::ZERO {
+            return Ok(false);
+        }
+        Ok(true)
+    }
+
+    /// Internal use only.
+    /// Converts the prime number to a sequence of `{0,1} ⊆ F`, and creates a
+    /// local replicated share.
+    fn local_secret_shared_bits_p<F: Field>(helper_role: Role) -> Vec<Replicated<F>> {
```

Comment:
There's probably some way to memoize this - so that it only needs to be computed once.

Reply:
Rust didn't support `memoize` by default, but found a crate for it. Added a TODO for potential optimization.